src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl

```diff
@@ -0,0 +1,171 @@
+- `batch_size::Int`
+- `update_freq::Int`
+- `update_step::Int`, count the step.
+- `rng::AbstractRNG`.
+"""
+mutable struct MADDPGManager{P<:DDPGPolicy, T<:AbstractTrajectory} <: AbstractPolicy
+    agents::Dict{<:Any, <:Agent{<:P, <:T}}
+    batch_size::Int
+    update_freq::Int
+    update_step::Int
+    rng::AbstractRNG
+end
+
+# for simultaneous game with a discrete action space.
+function (π::MADDPGManager)(env::AbstractEnv)
+    while current_player(env) == chance_player(env)
+        env |> legal_action_space |> rand |> env
+    end
+    Dict((player, ceil(agent.policy(env))) for (player, agent) in π.agents)
+end
+
+function (π::MADDPGManager)(::PreEpisodeStage, ::AbstractEnv)
+    for (_, agent) in π.agents
+        if length(agent.trajectory) > 0
+            pop!(agent.trajectory[:state])
+            pop!(agent.trajectory[:action])
+            if haskey(agent.trajectory, :legal_actions_mask)
+                pop!(agent.trajectory[:legal_actions_mask])
+            end
+        end
+    end
+end
+
+function (π::MADDPGManager)(::PreActStage, env::AbstractEnv, actions)
+    # update each agent's trajectory
+    for (player, agent) in π.agents
+        push!(agent.trajectory[:state], state(env, player))
+        push!(agent.trajectory[:action], actions[player])
+        if haskey(agent.trajectory, :legal_actions_mask)
+            lasm = legal_action_space_mask(env, player)
+            push!(agent.trajectory[:legal_actions_mask], lasm)
+        end
+    end
+    
+    # update policy
+    update!(π)
+end
+
+function (π::MADDPGManager)(::PostActStage, env::AbstractEnv)
+    for (player, agent) in π.agents
+        push!(agent.trajectory[:reward], reward(env, player))
+        push!(agent.trajectory[:terminal], is_terminated(env))
+    end
+end
+
+function (π::MADDPGManager)(::PostEpisodeStage, env::AbstractEnv)
+    # collect state and dummy action to each agent's trajectory
+    for (player, agent) in π.agents
+        push!(agent.trajectory[:state], state(env, player))
+        push!(agent.trajectory[:action], rand(action_space(env)))
+        if haskey(agent.trajectory, :legal_actions_mask)
+            lasm = legal_action_space_mask(env, player)
+            push!(agent.trajectory[:legal_actions_mask], lasm)
+        end
+    end
+
+    # update policy
+    update!(π)
+end
+
+# update policy
+function RLBase.update!(π::MADDPGManager)
+    π.update_step += 1
+    π.update_step % π.update_freq == 0 || return
+
+    for (_, agent) in π.agents
+        length(agent.trajectory) > agent.policy.update_after || return
+        length(agent.trajectory) > π.batch_size || return
+    end
+    
+    # get training data
+    temp_player = rand(keys(π.agents))
+    t = π.agents[temp_player].trajectory
+    inds = rand(π.rng, 1:length(t), π.batch_size)
+    batches = Dict((player, RLCore.fetch!(BatchSampler{SARTS}(π.batch_size), agent.trajectory, inds)) 
```

Comment:
The hardcoded `SARTS` will make the algorithm work only on environments of `MINIMAL_ACTION_SET`.